internal/slave/service.go

```diff
@@ -20,38 +25,58 @@ type DKVService interface {
+	// Applicable when replication requests are erroring out due to an issue with master / slave
+	MaxActiveReplElapsed uint64
+	// Listener address of the master node
+	ReplMasterAddr string
+	// Temporary flag to disable automatic master discovery until https://github.com/flipkart-incubator/dkv/issues/82 is fixed
+	// The above issue causes replication issues during master switch due to inconsistent change numbers
+	DisableAutoMasterDisc bool
 }
 
-// TODO: check if this needs to be exposed as a flag
-const maxNumChangesRepl = 10000
+type slaveService struct {
+	store        storage.KVStore
+	ca           storage.ChangeApplier
+	lg           *zap.Logger
+	statsCli     stats.Client
+	replCli      *ctl.DKVClient // can be nil when trying to find a master to replicate from
+	replTckr     *time.Ticker
+	regionInfo   *serverpb.RegionInfo
+	replStop     chan struct{}
+	replLag      uint64
+	fromChngNum  uint64
+	lastReplTime uint64
+	replConfig   *ReplicationConfig
+	clusterInfo  discovery.ClusterInfoGetter
+	isClosed     bool
+}
 
 // NewService creates a slave DKVService that periodically polls
 // for changes from master node and replicates them onto its local
 // storage. As a result, it forbids changes to this local storage
 // through any of the other key value mutators.
-func NewService(store storage.KVStore, ca storage.ChangeApplier, replCli *ctl.DKVClient, replPollInterval time.Duration,
-	lgr *zap.Logger, statsCli stats.Client) (DKVService, error) {
-	if replPollInterval == 0 || replCli == nil || store == nil || ca == nil {
-		return nil, errors.New("invalid args - params `store`, `ca`, `replCli` and `replPollInterval` are all mandatory")
+func NewService(store storage.KVStore, ca storage.ChangeApplier, lgr *zap.Logger,
+	statsCli stats.Client, regionInfo *serverpb.RegionInfo, replConf *ReplicationConfig, clusterInfo discovery.ClusterInfoGetter) (DKVService, error) {
+	if store == nil || ca == nil {
+		return nil, errors.New("invalid args - params `store`, `ca` and `replPollInterval` are all mandatory")
 	}
-	return newSlaveService(store, ca, replCli, replPollInterval, lgr, statsCli), nil
+	return newSlaveService(store, ca, lgr, statsCli, regionInfo, replConf, clusterInfo), nil
 }
 
-func newSlaveService(store storage.KVStore, ca storage.ChangeApplier, replCli *ctl.DKVClient, pollInterval time.Duration,
-	lgr *zap.Logger, statsCli stats.Client) *slaveService {
-	ss := &slaveService{store: store, ca: ca, replCli: replCli, lg: lgr, statsCli: statsCli}
-	ss.startReplication(pollInterval)
+func newSlaveService(store storage.KVStore, ca storage.ChangeApplier, lgr *zap.Logger,
+	statsCli stats.Client, info *serverpb.RegionInfo, replConf *ReplicationConfig, clusterInfo discovery.ClusterInfoGetter) *slaveService {
+	ss := &slaveService{store: store, ca: ca, lg: lgr, statsCli: statsCli,
+		regionInfo: info, replConfig: replConf, clusterInfo: clusterInfo}
+	ss.findAndConnectToMaster()
```

Comment:
We should maybe block/fail/retry-in-sync in case this connection attempt fails.

Reply:
not blocking as retry will occur automatically as part of replication attempt. so this enables slave to be started in case service discovery cluster is down and auto connect to master when service discovery cluster is available

Reply:
But in that case, this slave should not serve any requests, as the node may have failed to sync any/all data at all.

Reply:
slave won't server requests as status would be inactive as lastReplTime would not be set.